File: count-binary-substrings.ts
{
    function countBinarySubstrings(s: string) {
        let res = 0, prev = 0, curr = 1;
     
        for (let i = 1; i < s.length; i++) {
            if (s[i - 1] !== s[i]) {
                res += Math.min(prev, curr);
                prev = curr;
                curr = 1;
            } else {
                curr += 1;
            }
        }
     
        return res + Math.min(prev, curr);
    }
}

File: count-binary-substrings.rs
pub fn count_binary_substrings(s: String) -> i32 {
    let mut lk = 0;
    let mut rk = 1;
    let mut ans = 0;
    for (a, b) in s.chars().zip(s.chars().skip(1)) {
        if a != b {
            ans += lk.min(rk);
            lk = rk;
            rk = 0;
        }
        rk += 1;
    }
    ans += lk.min(rk);
    ans
}
